Init manager checking unready targets with target-aware watchers

ASOPVII opened the pull request:
Linked Issue: https://github.com/envoyproxy/envoy/issues/11963

**Description**: The current init manager only knows how many registered targets are not ready via its ```count_``` field. This pull request mainly helps init manager to check which specific targets are not ready. 

Key idea: Init manager stores the **unready_target_name:count** key-value pair in a hash map to check which registered targets are not ready. Enable passing target name into watcher's callback function.

1. When a new target is added into the targets list, increase the occurrence of the target's name in the hash map ```++target_name_count_[target_name]```.

2. When a target is ready, it informs the manager's watcher and finally, the watcher notifies the manager(that's why we need to pass a ```string_view name``` parameter to the manager's callback). And decrease the count of the target's name by 1.
 ```--target_name_count_[target_name]```.

**Most recent update**:
Wrapped the old ReadyFn type function into the new ```TargetAwareReadyFn```. Now ```WatcherImpl``` can take both types of function in its Ctor.

~Split WatcherImpl into two classes, the old one and a new ```TargetAwareWatcherImpl```. As we now limit the ```watcher_``` inside ```ManagerImpl``` to receive a ```string_view``` parameter, we declare it to be of ```TargetAwareWatcherImpl```.~

In the current code base, watchers are basically constructed with ```std::function<void()>```. In order to adopt the new feature(pass a ```string_view``` parameter to the manager's callback, I wrote a new constructor for ```WatcherImpl``` with ```std::function<void(absl::string_view)>``` parameter. In addition, I added a ```onTargetReadySendTargetName(string_view)``` adopted from the original callback ```onTargetReady()```. So now both types of callbacks are supported. I've also updated code in the constructor of ```ManagerImpl```, and now the ```watcher_``` of a manager is constructed with the new type of function(with string_view parameter). Therefore, the manager will always get target_name from the watcher's callback.

**Followup**: Unready targets config dumps

Risk Level: Low

The diagram below illustrates the outline of this PR
![envoy_init](https://user-images.githubusercontent.com/23137216/88350522-36244300-cd08-11ea-884b-2c1437848c2b.png)

lambdai commented on source/common/init/manager_impl.cc:
```diff
@@ -51,18 +54,34 @@ void ManagerImpl::initialize(const Watcher& watcher) {
     for (const auto& target_handle : target_handles_) {
       if (!target_handle->initialize(watcher_)) {
-        onTargetReady();
+        onTargetReady(target_names_[target_index]);
```
If we expose the target_name from TargetHandle(the only implementation contains the name anyway), 
the `target_names_` is not needed then. WDTY @mergeconflict

mergeconflict replied:
Agree, that will avoid duplicate memory usage.

lambdai commented on source/common/init/manager_impl.cc:
```diff
@@ -51,18 +54,34 @@ void ManagerImpl::initialize(const Watcher& watcher) {
 }
 
-void ManagerImpl::onTargetReady() {
+// For init manager to query unready targets
```
Period.

lambdai commented on source/common/init/manager_impl.cc:
```diff
@@ -51,18 +54,34 @@ void ManagerImpl::initialize(const Watcher& watcher) {
+void ManagerImpl::checkUnreadyTargets() {
+  ENVOY_LOG(debug, "QUERY: List of unready targets..................");
+  for(const auto &unready_target : target_names_count_) {
+    ENVOY_LOG(debug, "{} x {}", unready_target.first, unready_target.second);
```
I was hoping Manager could provide a function to dump to protobuf message and eventually used by admin handler.
It's no harm to log though.

lambdai commented:
CC @jplevyak

mergeconflict commented on include/envoy/runtime/runtime.h:
```diff
@@ -266,7 +266,7 @@ class Loader {
   virtual ~Loader() = default;
 
-  using ReadyCallback = std::function<void()>;
+  using ReadyCallback = std::function<void(const std::string)>;
```
Passing `const std::string` is very unusual c++; it makes a copy that can't be modified in the function. Prefer passing `absl::string_view`, and don't use `const` for value params. This goes for all instances below.

mergeconflict commented on source/common/init/manager_impl.h:
```diff
@@ -56,6 +61,12 @@ class ManagerImpl : public Manager, Logger::Loggable<Logger::Id::init> {
+  std::vector<std::string> target_names_;
+
+  // Count of target_name
+  std::unordered_map<std::string, uint32_t> target_names_count_;
```
I don't understand the purpose of this. Why would somebody want to have multiple init targets registered with the same name?

lambdai replied:
I think no one would intentionally do it :)
But I would also assume duplicated names is allowed somehow.

mergeconflict replied:
It's probably a bug if somebody does this. I think we should assert that init target names are unique within an init manager, and change this from a map to a set.

Also, actually, it shouldn't be a `std::unordered_`anything, it should be an `absl::flat_hash_set` :)

mergeconflict commented on source/common/init/manager_impl.cc:
```diff
@@ -51,18 +54,34 @@ void ManagerImpl::initialize(const Watcher& watcher) {
 
-void ManagerImpl::onTargetReady() {
+// For init manager to query unready targets
+void ManagerImpl::checkUnreadyTargets() {
```
This doesn't seem to be called from anywhere other than tests. Where do you expect to call this from?

lambdai replied:
Eventually the dumped info should be returned and will be driven by admin handler when iterator over init manager owners.
https://github.com/envoyproxy/envoy/issues/11963

I think checkUnreadyTargets() should be erased or for debug only.

ASOPVII commented:
Thank you guys for the great reviews~ 'checkUnreadyTargets()' is just a basic function to check if the method could work.  I am working on dumping the info to protobuf message and finally be used by admin handler.

junr03 commented:
/wait

mergeconflict commented on source/common/init/manager_impl.cc:
```diff
@@ -53,16 +55,30 @@ void ManagerImpl::initialize(const Watcher& watcher) {
-void ManagerImpl::onTargetReady() {
+// For init manager to query unready targets.
+void ManagerImpl::checkUnreadyTargets() {
+  ENVOY_LOG(debug, "QUERY: List of unready targets..................");
```
What does "query" mean in this context? Can this log entry just say `List of unready targets:`?

mergeconflict commented on source/common/init/manager_impl.cc:
```diff
@@ -53,16 +55,30 @@ void ManagerImpl::initialize(const Watcher& watcher) {
+  for (const auto& unready_target : target_names_count_) {
+    ENVOY_LOG(debug, "{} x {}", unready_target.first, unready_target.second);
+  }
+  ENVOY_LOG(debug, "QUERY ENDS......................................");
```
Remove

mergeconflict commented on source/server/server.cc:
```diff
@@ -506,7 +506,8 @@ void InstanceImpl::initialize(const Options& options,
   // If RTDS was not configured the `onRuntimeReady` callback is immediately invoked.
-  Runtime::LoaderSingleton::get().startRtdsSubscriptions([this]() { onRuntimeReady(); });
+  Runtime::LoaderSingleton::get().startRtdsSubscriptions(
+      [this](absl::string_view) { onRuntimeReady(); });
```
One thing I notice looking at this PR is that most callbacks look like this now:
```c++
[this](absl::string_view) { /* ignore the string_view param! */ }
```
(I'm randomly leaving the comment here on this line, but it applies everywhere). This feels like a code smell; the name is only used internally in the init manager for debugging. Could there be a different way to implement this that doesn't require leaking names into unrelated code like this?

ASOPVII replied:
That's what's annoying me these days too because I need to pass the target name info into the callback so that the manager could update its status.

mergeconflict commented on test/common/init/manager_impl_test.cc:
```diff
@@ -22,9 +22,11 @@ TEST(InitManagerImplTest, AddImmediateTargetsWhenUninitialized) {
 
   ExpectableTargetImpl t1("t1");
   m.add(t1);
+  m.checkUnreadyTargets();
```
I don't think that adding these checks in the unit tests does anything. That is, the unit test isn't testing the content of the debug messages, right?

ASOPVII replied:
Yes, agreed. I will remove them in the next commit.

ASOPVII commented:
> LGTM
> 
> Could you link the issue or address the follow up in description?

Sure, updated.

ASOPVII commented:
**help wanted**:
In my previous design, a string parameter is passed into the manager's callback "onTargetReady()". The purpose of this additional parameter is to tell the init manager which target is ready. As @mergeconflict mentioned, this could be a code smell. 

I am trying to come up with a better implementation. One idea is to create additional watchers for each target. In this way, the manager will know targets through their corresponding watcher. But this is just a simple, basic idea, not sure if it is ok to create many watchers. 

Suggestions are welcome, please comment below if you have any idea.

ASOPVII commented:
**Help Wanted**:
I failed Linux-x64 release pre-submit test with two test cases: 
```IpVersions/ClusterMemoryTestRunner.MemoryLargeClusterSizeWithFakeSymbolTable/IPv4, where GetParam() = 4-byte object <00-00 00-00>``` and ```IpVersions/ClusterMemoryTestRunner.MemoryLargeClusterSizeWithRealSymbolTable/IPv4, where GetParam() = 4-byte object 
<00-00 00-00>```

Anyone knows the reason of this failure?

mergeconflict commented on source/common/init/watcher_impl.h:
```diff
@@ -24,23 +25,35 @@ using ReadyFn = std::function<void()>;
+  // Ctor with std::function<void()> callback function.
   WatcherHandleImpl(absl::string_view handle_name, absl::string_view name,
                     std::weak_ptr<ReadyFn> fn);
 
+  // Ctor with std::function<void(absl::string_view)> callback function.
+  WatcherHandleImpl(absl::string_view handle_name, absl::string_view name,
+                    std::weak_ptr<ReadyFnSendName> fn);
+
 public:
-  // Init::WatcherHandle
+  // Init::WatcherHandle.
   bool ready() const override;
 
 private:
   // Name of the handle (either the name of the target calling the manager, or the name of the
-  // manager calling the client)
+  // manager calling the client).
   const std::string handle_name_;
 
-  // Name of the watcher (either the name of the manager, or the name of the client)
+  // Name of the watcher (either the name of the manager, or the name of the client).
   const std::string name_;
 
-  // The watcher's callback function, only called if the weak pointer can be "locked"
+  // When the function constructed is ReadyFnSendName, set this to true.
+  bool fn_send_name_;
+
+  // The watcher's callback function, only called if the weak pointer can be "locked".
   const std::weak_ptr<ReadyFn> fn_;
+
+  // The watcher's callback function, only called if the weak pointer can be "locked".
+  const std::weak_ptr<ReadyFnSendName> fn_sn_;
```
Not sure I agree with this design. It basically says "if you want this watcher to send a name, construct me with a `ReadyFnSendName`, and if not, ehhh whatever." That means you'll have inconsistent usage across the codebase. I think you want your design to enforce consistency so that the information is always there when you need it.

ASOPVII replied:
Thanks for the advice. Here the two options of ctors of the watcherImpl are mainly designed to support the old codes and my new codes. 

But the main idea is that the manager could get the information of target names. As in my current code, the manager is constructed with a watcher(with [](string_view){} as [function](https://github.com/ASOPVII/envoy/blob/init-manager-query-unready-targets/source/common/init/manager_impl.cc#L11-L15)), and it is when the target calls its initialize that the watcher gets a name, (code [link](https://github.com/ASOPVII/envoy/blob/init-manager-query-unready-targets/source/common/init/target_impl.cc#L16)). Therefore, even the old code does not pass a parameter, it could be regarded as an empty string_view, which does not influence the old logic. In addition, if people want to use this feature, they could try to add string_view parameter to the ctor of watcher, so that the manager could get real target names. I think this

ASOPVII replied:
The best way to understand this pr is that it not only maintains the old version of watcher ctor, but also supports new ctor, which provides the manager a method to know target names. Even though that in the old code base, watchers are most init with ```[](){}```, it does only have effect on the old tests. Since now the manager has been changed to receive a new type of function, the manager will always get the information of target names. I think the code is not only consistent with the old version, but helps to provide the features. Please notice that managerImpl ctor is the main point. WatcherImpl ctors only help to support both versions(most old test case where watchers are init with ```[](){}```.)

ASOPVII replied:
> Not sure I agree with this design. It basically says "if you want this watcher to send a name, construct me with a `ReadyFnSendName`, and if not, ehhh whatever." That means you'll have inconsistent usage across the codebase. I think you want your design to enforce consistency so that the information is always there when you need it.

Thanks for the review. The point is not focused on WatcherImpl. Please look at the impl of ManagerImpl, now the watcher_ of the ManagerImpl is constructed with ```[](string_view){}```, so it should always get the target names. I believe there is no inconsistency, even though the watchers in old tests are mostly constructed with ```[](){}```, it does not influence the managerImpl.

lizan replied:
I don't like this either. I guess you would be able to roll `ReadyFnSendName` into `ReadyFn` by capturing the name in the lambda, and the rest of WatcherHandle changes won't be necessary.

ASOPVII commented:
@mergeconflict You can actually regard the two ctors as two watcherImpl classes. The ```watcher_``` inside ```ManagerImpl``` is now implemented as an ```InternalWatcherImpl``` which has a ```string_view``` parameter in the callback function. This helps the init manager to get target_name information from the watcher_ callback. And most other watchers defined independently in the project remain the same as the old implementation, though you can now construct them in any one of the two ways. I only changed the internal ```watcher_``` of init manager to be constructed with the new function, which does not influence any other place throughout the project. If you still can not approve that implementation, I can just write a new class ```InternalWatcherImpl``` and defines the watcher_ inside ManagerImpl as a instance of that new class, while keeping the old WatcherImpl unchained, although it does the same things as the current one does, one class with 2 ctors or just 2 classes.

mergeconflict commented:
Hi, sorry for the slow reply. I'm out on parental leave, so I'm sort of mostly unavailable...

> @mergeconflict You can actually regard the two ctors as two watcherImpl classes.

True, but even with two distinct watcher implementations, my point is the same: I think having two impls defeats your purpose because people will just use whichever one is easiest and you'll have incomplete data on which targets are ready.

I don't want to hold up your work though, you're trying to do something useful :) Maybe the solution is just to document clearly which constructor should be used in which situations. Have a chat with @htuch or @lizan and see if they have opinions?

ASOPVII commented:
> Hi, sorry for the slow reply. I'm out on parental leave, so I'm sort of mostly unavailable...
> 
> > @mergeconflict You can actually regard the two ctors as two watcherImpl classes.
> 
> True, but even with two distinct watcher implementations, my point is the same: I think having two impls defeats your purpose because people will just use whichever one is easiest and you'll have incomplete data on which targets are ready.
> 
> I don't want to hold up your work though, you're trying to do something useful :) Maybe the solution is just to document clearly which constructor should be used in which situations. Have a chat with @htuch or @lizan and see if they have opinions?

Great idea. I will absolutely write more comments about the usage of these two ctors. But I have to say that in this implementation, init managers will always know which targets are ready, since no matter in which way you define a watcher outside somewhere, the watcher inside the ManagerImpl is now always constructed with an additional string_view parameter. So there is no need to worry about incomplete data. As long as the init manager adds a target, it now keeps its name in the hash map and the internal watcher_  of this manager will notify the target_name in the future when the target is ready.

ASOPVII commented:
@mergeconflict More comments have been added to describe the usage of WatcherImpl' s Ctor. I've also made a diagram(post in the top of this PR) to illustrate this work. Please help review~

ASOPVII commented on api/envoy/admin/v3/config_dump.proto:
```diff
@@ -381,3 +381,19 @@ message EndpointsConfigDump {
 }
+
+// Message of unready targets information of listener's init manager.
+message ListenerUnreadyTargetsConfigDump {
```
Great idea! Thanks for the review. Updated

ASOPVII replied:
However, for the second part, I think include name into lambda function seems not feasible, since the internal watcher of init manager is defined at Ctor of ManagerImpl and could not pass a string name there. 

What about another solution: since now I have changed the internal watcher of init manager to be fixed to the Ctor with ```std::function<void(string)>```, and for the feature of init manager 'check unready targets', we have to make that ```watcher_``` to be constructed with that. Is it possible that I just split them into two types of watchers. One for internal watcher of init manager(which must be ```<void(string)>```, and the other one type of watcher for non-initmanager, which mostly is for target.

lizan replied:
I think you can do the other way at least? i.e. wrapping ReadyFn into ReadyFnSendName (I don't think this type is good btw).

i.e.
```
ReadyFn fn;
ReadyFnSendName wrapped = [fn](absl::string_view) -> void { fn(); };
```

ASOPVII replied:
Yes, I get the idea of wrapping fn into lambda. But the problem is, if I want to return the wrapped function, then watcher will become ```watcher_(name, [this] () { wrapped; return wrapped; })```. I will need to change ```ReadyFn=std::function<void()>``` into ```ReadyFn=std::function<std::function<void(string)>()>```.

ASOPVII replied:
```C++
using ReadyFnSendName = std::function<void(absl::string_view)>;
using ReadyFn = std::function<ReadyFnSendName()>;

watcher_(name_, [this]() {
  ReadyFnSendName fn = [this](absl::string_view name) {
    onTargetReadySendTargetName(name);
  };
  return fn;
}) {}


auto locked_fn(fn_.lock());
if (locked_fn) {
  // If we can "lock" a shared pointer to the watcher's callback function, call it.
  ENVOY_LOG(debug, "{} initialized, notifying {}", handle_name_, name_);
  auto send_name_fn = (*locked_fn)();
  send_name_fn(handle_name_);
  // (*locked_fn)();
  return true;
}
```

ASOPVII replied:
This is what wrapping Fn looks like. Since ```watcher_```'s function can not contain any parameter, we have to change its return type from ```void``` to ```ReadyFnSendName```.

ASOPVII replied:
> I think you can do the other way at least? i.e. wrapping ReadyFn into ReadyFnSendName (I don't think this type name is good btw).
> 
> i.e.
> 
> ```
> ReadyFn fn;
> ReadyFnSendName wrapped = [fn](absl::string_view) -> void { fn(); };
> ```

The wrapping seems to not solve this problem, because the ```onTargetReady()``` now receives a string parameter. And the major part is inside ```fn```, which is passed from manager. The wrapping still could not change the truth the  ```fn``` needs to receive string.

ASOPVII replied:
> I think you can do the other way at least? i.e. wrapping ReadyFn into ReadyFnSendName (I don't think this type name is good btw).
> i.e.
> 
> ```
> ReadyFn fn;
> ReadyFnSendName wrapped = [fn](absl::string_view) -> void { fn(); };
> ```

Now that the callback function ```onTargetReady```  will receive a string parameter, please tell me how can I wrap this function a ReadyFn then call it while passing a string parameter.

lambdai commented on source/server/admin/admin.cc:
```diff
@@ -456,15 +467,26 @@ Http::Code AdminImpl::handlerClusters(absl::string_view url,
-                                   const absl::optional<std::string>& mask,
-                                   bool include_eds) const {
+                                   const absl::optional<std::string>& mask, bool include_eds,
+                                   bool include_active_listener_unready_targets,
+                                   bool include_warming_listener_unready_targets) const {
```
It's over complicated. The `not ready listener in active list` is an implementation details.
User would only care about the the unready listeners regardless they are in active list or warm list.

ASOPVII replied:
Will merge them together.

lambdai commented on source/server/admin/admin.cc:
```diff
@@ -619,24 +652,59 @@ ProtobufTypes::MessagePtr AdminImpl::dumpEndpointConfigs() const {
   return endpoint_config_dump;
 }
 
+ProtobufTypes::MessagePtr AdminImpl::dumpUnreadyTargetsConfigs(bool dump_active_listeners) const {
```
We only need dump all targets. The targets could comes from
1. `active listeners`, if workers are not started yet.
2.  `warm listeners`, if workers are started.

ASOPVII commented on source/common/init/manager_impl.cc:
```diff
@@ -70,6 +82,22 @@ void ManagerImpl::onTargetReady() {
 
+void ManagerImpl::onTargetReadySendTargetName(absl::string_view target_name) {
+  // If there are no remaining targets and one mysteriously calls us back, this manager is haunted.
+  ASSERT(count_ != 0, fmt::format("{} called back by target after initialization complete"));
```
This part is copied from original ```onTargetReady()```. Added the ```target_name``` in this part, but what about the ```{}``` inside original ```onTargetReady()```

ASOPVII commented on source/common/init/watcher_impl.cc:
```diff
@@ -29,10 +29,42 @@ WatcherImpl::~WatcherImpl() { ENVOY_LOG(debug, "{} destroyed", name_); }
+    return true;
+  } else {
+    // If not, the watcher was already destroyed.
+    ENVOY_LOG(debug, "{} initialized, but can't notify {} (unavailable)", handle_name_, name_);
```
Same as above, this part comes from ```WatcherHandlerImpl```, which already exists.

ASOPVII commented on source/common/init/manager_impl.h:
```diff
@@ -35,27 +37,39 @@ class ManagerImpl : public Manager, Logger::Loggable<Logger::Id::init> {
-  // Watcher to receive ready notifications from each target
-  const WatcherImpl watcher_;
+  // Watcher to receive ready notifications from each target.
+  const TargetAwareWatcherImpl watcher_;
```
Now the ```watcher_``` inside manager is defined as a ```TargetAwareWatcher```

lizan commented on source/common/init/watcher_impl.h:
```diff
@@ -51,23 +54,86 @@ class WatcherHandleImpl : public WatcherHandle, Logger::Loggable<Logger::Id::ini
+
+private:
+  // Name of the handle (either the name of the target calling the manager, or the name of the
+  // manager calling the client).
+  const std::string handle_name_;
+
+  // Name of the watcher (either the name of the manager, or the name of the client).
+  const std::string name_;
+
+  // The watcher's callback function, only called if the weak pointer can be "locked".
+  const std::weak_ptr<TargetAwareReadyFn> fn_;
+};
+
+/**
+ * A TargetAwareWatcherImpl is a WatcherImpl which is specially designed for init manager's internal
+ * watcher_. This watcher_ will monitor all the targets this init manager has added. The callback
+ * function fn has a absl::string_view parameter to pass the target name to init manager.
+ */
+class TargetAwareWatcherImpl : public Watcher, Logger::Loggable<Logger::Id::init> {
+public:
+  /**
+   * @param name a human-readable watcher name, for logging / debugging.
+   * @param fn a callback function to invoke when `ready` is called on the handle.
+   */
+  TargetAwareWatcherImpl(absl::string_view name, TargetAwareReadyFn fn);
+  ~TargetAwareWatcherImpl() override;
+
+  // Init::Watcher.
+  absl::string_view name() const override;
+  WatcherHandlePtr createHandle(absl::string_view handle_name) const override;
+
+private:
+  // Human-readable name for logging.
+  const std::string name_;
+
+  // The callback function with target_name parameter, called via WatcherHandleImpl by either the
+  // target or the manager.
+  const std::shared_ptr<TargetAwareReadyFn> fn_;
+};
+
```
Why do you still need these classes? `WatcherImpl`/`WatcherHandleImpl` can be just like these with default empty names?

ASOPVII replied:
Sorry, do you mean that change ReadyFn to ```std::function<void(absl::string_view)>```?

lizan replied:
Yes, why not?

ASOPVII replied:
The problem is, in the current code base, most watchers are defined as ```[]() { }``` and the absl::string_view parameter will not automatically transform an empty parameter into a empty ```string_view```. I have to change all those definitions from ```[](){} ```into ```[](""){}```

lizan replied:
You can overload constructor to take [](){} and wrap with empty string?

ASOPVII replied:
Several attempts I have tried on this project:
1. Directly change ```ReadyFn``` from ```std::function<void()>``` into  ```std::function<void(absl::string_view)>```. Turns out that due to the most definitions in the current code base are ```[](){}```(Like hundreds of definitions in tests and other places). There will be errors unless I change all of them from ```[](){}``` into ```[](""){}```. This solution is not good because there are too many current definitions to update and for potential wrong definition in the future usage.
2. Add additional Ctor, which could receive both ```ReadyFn```  and ```std::function<void(string_view)>```. Any one of them could work. ReadyFn works well for existing definitions and the new one support the new feature(the internal watcher of manager).

ASOPVII replied:
> Yes, why not?

Since we need to send name to manager, the ```onTargetReady``` callback has to receive a string parameter now. So the callback passed to watcher has to be called somewhere with a string name. I thought about template and lambda, but that will not change the truth that watcher needs to pass string and call ```onTargetReady(string)```.

ASOPVII replied:
> You can overload constructor to take {} and wrap with empty string?

I am not sure if that could work as I need to make the function shared. But I will try.

ASOPVII replied:
> You can overload constructor to take {} and wrap with empty string?

I think the two types of Ctor in my previous version is what is done here. Basically, the wrapping looks like ```fn_( [](""){std::move(fn); ) )```. I don't see the difference of using two Ctors and wrapping one into the other.

ASOPVII replied:
Lets assume ```A=std::function<void()>``` and ```B=std::function<void(absl::string_view)>```
Existing definition of watcher receives: ```[](){}``` as function. This is class A. It is able to wrap A into B:
````
B fn_;
Ctor(A) {
    fn_ = []("") { A }
}.
````
But finally, watchers Ctor with A naturally does not need to pass string into A, which means this watcher basically is higher level watcher which does not belong to any manager and it does not need to send name. It's meaningless wrapping its A into B. 
So I suppose it is more clear to make them different(with 2 Ctors or 2 separate Classes).

ASOPVII replied:
> You can overload constructor to take {} and wrap with empty string?

Given that watcher is now defined as 
```c++
watcher(name, [](string name) {
    onTargetReady(name);
}
``` 
I think its impossible to roll this ```[](string){ f(string); }``` into a ```<void()>```.

ASOPVII commented on source/common/init/watcher_impl.cc:
```diff
@@ -4,34 +4,38 @@ namespace Envoy {
 }
 
 WatcherImpl::WatcherImpl(absl::string_view name, ReadyFn fn)
-    : name_(name), fn_(std::make_shared<ReadyFn>(std::move(fn))) {}
+    : name_(name), fn_(std::make_shared<TargetAwareReadyFn>(
+                       [callback = std::move(fn)](absl::string_view) { callback(); })) {}
+
+WatcherImpl::WatcherImpl(absl::string_view name, TargetAwareReadyFn fn)
+    : name_(name), fn_(std::make_shared<TargetAwareReadyFn>(std::move(fn))) {}
```
Anyway, I removed the separate class ```TargetAwareWatcher```. Now the ```WatcherImpl``` can take both types of function in its Ctor. And the old one ```ReadyFn``` will be wrapped into the new type ```TargetAwareReadyFn```. Really appreciate your patience reviewing my code~.

ASOPVII commented:
Comment previous trials:
1. Directly change ```ReadyFn``` from ```std::function<void()>``` into  ```std::function<void(absl::string_view)>```. Turns out that due to most definitions in the current code base are ```[](){}```(Like hundreds of definitions in tests and in some other hidden cases). There will be errors unless I change all of them from ```[](){}``` into ```[](""){}```. This solution is not good because there are too many current definitions to update and for potential wrong definition in the future usage.
2. Add additional Ctor, so that ```watcher``` could receive both ```ReadyFn```  and ```std::function<void(string_view)>```. Any one of them could work. ReadyFn works well for existing definitions and the new one support the new feature(the internal watcher of manager).
3. I just directly introduced a new watcher class 'TargetAwareWatcher', which is adopted from the original one.
4. Removed ```TargetAwareWatcher``` Class. Wrapped the received ```ReadyFn``` function into ```TargetAwareReadyFn```. 

There are basically two types of watcher. One is the watcher owned by the manager, we have now enabled it to send target names. 
The other type is other high level watchers which typically monitors the manager, which is the one passed into manager's Ctor.

lizan commented on source/common/init/manager_impl.h:
```diff
@@ -35,27 +37,42 @@ class ManagerImpl : public Manager, Logger::Loggable<Logger::Id::init> {
+
+  // Callback function with an additional target_name parameter, decrease unready targets count by
+  // 1, update target_names_count_ hash map.
+  void onTargetReadySendTargetName(absl::string_view target_name);
```
We can name this `onTargetReady(absl::string_view target_name);`

ASOPVII replied:
Updated.

lizan commented on source/common/init/manager_impl.cc:
```diff
@@ -70,6 +81,23 @@ void ManagerImpl::onTargetReady() {
+         fmt::format("{} called back by target after initialization complete", target_name));
+
+  // Decrease target_name count by 1.
+  if (--target_names_count_[target_name] == 0) {
```
add an `ASSERT` that target_names_count_ contains this entry.

ASOPVII replied:
Updated.

lizan commented on source/common/init/manager_impl.cc:
```diff
@@ -70,6 +81,23 @@ void ManagerImpl::onTargetReady() {
+  if (--target_names_count_[target_name] == 0) {
+    target_names_count_.erase(target_name);
+  }
+
+  // If there are no uninitialized targets remaining when called back by a target, that means it was
+  // the last. Signal `ready` to the handle we saved in `initialize`.
+  if (--count_ == 0) {
+    ready();
+  }
```
just call `onTargetReady()`?

ASOPVII replied:
Maybe we can just delete the original ```onTargetReady()``` since it we not be used any more?

lizan replied:
yep sgtm.

ASOPVII replied:
Updated.

lizan commented on source/common/init/watcher_impl.cc:
```diff
@@ -4,34 +4,38 @@ namespace Envoy {
 }
 
 WatcherImpl::WatcherImpl(absl::string_view name, ReadyFn fn)
-    : name_(name), fn_(std::make_shared<ReadyFn>(std::move(fn))) {}
+    : name_(name), fn_(std::make_shared<TargetAwareReadyFn>(
+                       [callback = std::move(fn)](absl::string_view) { callback(); })) {}
```
Do we know what targets are relying on this path? Ideally I'd like to see all init targets are with names at the end. If there's a lot, can you file an issue to track?

ASOPVII replied:
No, I don't think we can get the target names. Since the passed in callback is ```std::function<void()>```. And I suppose this type of watcher(Ctor with ```<void()>```) has not been given tasks to track the name.

lambdai replied:
https://github.com/envoyproxy/envoy/pull/12035#discussion_r453977229 
The initial discussion to respect no target name as argument is above.

We can either choose passing target name everywhere, which is done at the initial commits,
or we allow function doesn't careabout target name as current

lizan replied:
My question is what is using this without name. They are basically tech debts, no?

ASOPVII replied:
Watchers that are passed from higher level to a dynamic init manager is using this kind of Ctor. Callbacks of these watchers does not need to pass names. Commonly their callback is ```higher_level_target.ready()``` and ```ready()``` of these targets does not need names.

ASOPVII replied:
So that's why I proposed to separate watchers into 2 types, one for targets and the other for init manager. Currently only those for init managers will need to pass names to managers.

lizan replied:
Yeah, I'm fine leaving those usages and keep them separately, just want to know what are using them.

lizan commented on source/server/admin/admin.cc:
```diff
@@ -455,15 +461,19 @@ Http::Code AdminImpl::handlerClusters(absl::string_view url,
-                                   const absl::optional<std::string>& mask,
-                                   bool include_eds) const {
+                                   const absl::optional<std::string>& mask, bool include_eds,
+                                   bool include_listener_unready_targets) const {
```
Since the API is no longer listener specific, should we make this one generic as well? (through out this file)

ASOPVII replied:
Emm, but actually the current ```dumoUnreadyTargetsConfigs``` only dumps unready targets of (warming) listeners

lizan replied:
How about in long term? We're likely having more unready targets when we address ECDS / On-demand DS / asynchronous Filter initialization?

ASOPVII replied:
I am not sure how do we specify whose unready targets do we want. And how do we track their init manager in a general approach.

ASOPVII replied:
Updated. Added a general method:  ```dumpUnreadyTargetsConfig(optional::string mask)``` to use mask to decide what to dump.

ASOPVII commented:
> Can we split the init manager refactoring part to another PR?

But I don't think init manager's watcher 

> Can we split the init manager refactoring part to another PR?

You mean 1 PR for TargetAwareWatcher and 1 PR for manager's onTargetReady? SGTM

lizan commented:
You can put both of Init manager / target to a single PR, I meant splitting this PR to the following two:
- refactoring init manager / target / handler to optionally take names
- dump unready targets in admin endpoint

ASOPVII commented:
@lizan I've updated this PR and separated the config dump into another branch, please review the recent commit.

lizan commented on source/common/init/manager_impl.cc:
```diff
@@ -18,13 +21,15 @@ void ManagerImpl::add(const Target& target) {
   case State::Uninitialized:
     // If the manager isn't initialized yet, save the target handle to be initialized later.
     ENVOY_LOG(debug, "added {} to {}", target.name(), name_);
+    ++target_names_count_[target.name()];
```
This should be before `swtich`right below (`++count_;`)

ASOPVII replied:
I suppose we should not add target into the hash table when the state is ```Initialized```.

lizan replied:
Right, the statement after `case State::Initialized` is `ASSERT(false)` so it will crash the whole process anyway :)

ASOPVII replied:
Yeah, updated

lizan commented on source/common/init/watcher_impl.h:
```diff
@@ -24,23 +25,25 @@ using ReadyFn = std::function<void()>;
 private:
   friend class WatcherImpl;
+
+  // Ctor with std::function<void()> callback function.
```
comment doesn't match the function is with string_view.

lizan commented on include/envoy/init/target.h:
```diff
@@ -25,6 +25,11 @@ struct TargetHandle {
   virtual bool initialize(const Watcher& watcher) const PURE;
+
+  /**
+   * @return a human-readable target name, for logging / debugging.
```
This is not only for logging / debugging but for target tracking as well? Comment it it have to be unique (otherwise the ASSERT below can fail).

lizan commented:
Can you also update PR title and description?

ASOPVII commented:
> Can you also update PR title and description?

Sure, updated.

ASOPVII commented:
> Can you also update PR title and description?

I think everything is completed now~

lizan commented:
Thanks!
